### PR TITLE
Type Migration: introducing primitive functional interfaces

### PR DESCRIPTION
**Pull request by olgsm**
Hey!

We are a collaborative group of researchers from JetBrains Research and Oregon State University, and we are testing our data-driven [plugin](https://github.com/JetBrains-Research/data-driven-type-migration), which is based on the IntelliJ's [Type Migration](https://www.jetbrains.com/help/idea/type-migration.html) framework and adjusts it using custom structural-replace templates that express the adaptations required to perform the type changes.

I want to apply several type changes using it and open the PR, thus introducing primitive functional interfaces in order to prevent unnecessary boxing (like BooleanSupplier instead of Supplier<Boolean>, OptionalInt instead of Optional<Integer>, Predicate<T> instead of Function<T, Boolean>, etc.), since it can affect the performance of the code (Effective Java, Items 44, 61). 

This patch is made automatically by the plugin, so it would help us a lot to evaluate the usefulness of our approach!

Thank you in advance!

### CODE REVIEW

**Comment by henningandersen** [comment]
@elasticmachine test this please

**Review by henningandersen — CHANGES_REQUESTED** [review]
This looks mostly good, but would like to see one comment addressed before merging.

File: server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java [hunk 1]
```diff
@@ -378,15 +377,15 @@ public DiscoveryNode resolveNode(String node) {
                                 // the role is not a data role, we require an exact match (e.g., ingest)
                                 predicate = s -> s.contains(role);
                             }
-                            final Function<String, Boolean> mutation;
+                            final Predicate<String> mutation;
```
[the remark this answers is not in the document]
Comment (SmirnovOleg) [reply server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java]: Fixed

**Comment by SmirnovOleg** [comment]
@henningandersen @ChrisHegarty Thank you for such quick feedback!
Hope that I've fixed all the comments in the last commits!

File: x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/action/TransportDeleteExpiredDataActionTests.java [hunk 2]
```diff
@@ -50,11 +50,11 @@
     private static class DummyDataRemover implements MlDataRemover {
 
         public void remove(
-            float requestsPerSec,
-            ActionListener<Boolean> listener,
-            Supplier<Boolean> isTimedOutSupplier
+                float requestsPerSec,
+                ActionListener<Boolean> listener,
+                BooleanSupplier isTimedOutSupplier
```
Comment (henningandersen) [inline x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/action/TransportDeleteExpiredDataActionTests.java:55]: I think this indentation change does not match our conventions, can you remove the extra indentation?

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/retention/ExpiredForecastsRemover.java [hunk 3]
```diff
@@ -100,18 +100,18 @@ public void remove(float requestsPerSec, ActionListener<Boolean> listener, Suppl
...
 
     private void deleteForecasts(
-        SearchResponse searchResponse,
-        float requestsPerSec,
-        ActionListener<Boolean> listener,
-        Supplier<Boolean> isTimedOutSupplier
+            SearchResponse searchResponse,
+            float requestsPerSec,
+            ActionListener<Boolean> listener,
+            BooleanSupplier isTimedOutSupplier
```
Comment (henningandersen) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/retention/ExpiredForecastsRemover.java:106]: I think this indentation change does not match our conventions, can you remove the extra indentation?

**Review by henningandersen — CHANGES_REQUESTED** [review]
Thanks for addressing the initial feedback. I missed a few whitespace-nits initially, hope you do not mind resolving those too.

**Comment by SmirnovOleg** [comment]
@henningandersen  
Done!

**Comment by henningandersen** [comment]
@elasticmachine test this please

**Comment by henningandersen** [comment]
@elasticmachine update branch

**Comment by henningandersen** [comment]
@elasticmachine test this please

**Review by henningandersen — APPROVED** [review]
LGTM, thanks for your contribution.

**Pull request merged**